packages/labs/ssr-react/src/lib/node/render-shadow-contents.ts

```diff
@@ -0,0 +1,81 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {LitElementRenderer} from '@lit-labs/ssr/lib/lit-element-renderer.js';
+import {getElementRenderer} from '@lit-labs/ssr/lib/element-renderer.js';
+import type {RenderInfo} from '@lit-labs/ssr';
+
+const reservedReactProperties = new Set([
+  'children',
+  'localName',
+  'ref',
+  'style',
+  'className',
+]);
+
+const attributesToProps = (attrs: NamedNodeMap) => {
+  const props: {[index: string]: string} = {};
+  for (let i = 0; i < attrs.length; i++) {
+    const attr = attrs[i];
+    props[attr.name] = attr.value;
+  }
+  return props;
+};
+
+/**
+ * Renders the shadow contents of the provided custom element type with props.
+ * Should only be called in server environments.
+ */
+export const renderShadowContents = (type: string, props: {} | null) => {
+  const renderInfo: RenderInfo = {
+    elementRenderers: [LitElementRenderer],
+    customElementInstanceStack: [],
+    customElementHostStack: [],
+    deferHydration: false,
+  };
+
+  const renderer = getElementRenderer(renderInfo, type);
+
+  if (renderer.element !== undefined && props != null) {
+    for (const [k, v] of Object.entries(props)) {
+      // Reserved React Props do not need to be set on the element
+      if (reservedReactProperties.has(k)) {
+        continue;
+      }
+
+      if (k in renderer.element) {
+        renderer.setProperty(k, v);
+      } else {
+        renderer.setAttribute(k, String(v));
+      }
+    }
+  }
+
+  renderer.connectedCallback();
+
+  renderInfo.customElementInstanceStack.push(renderer);
+
+  const shadowContents = renderer.renderShadow(renderInfo);
+
+  // elementAttributes will be provided to React as props for the host element
+  // for properly rendering reflected attributes
+  const elementAttributes =
+    renderer.element !== undefined
+      ? attributesToProps(renderer.element.attributes)
+      : {};
+
+  const {mode = 'open', delegatesFocus} = renderer.shadowRootOptions;
+  const templateAttributes = {
+    shadowroot: mode,
```

Comment:
Change to `shadowrootmode: mode` for compatibility with Safari and streaming SSR. We need to update our polyfill too.

Reply:
Should we keep the existing `shadowroot: mode` as well for compatibility with older Chrome versions? Per Elliott's testing, it seems fine to have both present and will have the streaming behavior for supported browsers. The streaming DSD feature that also renames the attribute is estimated to ship in Chrome 111 https://chromestatus.com/feature/5161240576393216

Reply:
I think we can launch with just the new attribute considering that we have no existing users. We'd have to deprecate the old attribute right away anyway.

Reply:
I'm hesitant on launching with an attribute that doesn't work on any currently shipping stable browsers. Our core labs/ssr package also does both which means nested custom elements will get both attributes. We should match that behavior and deprecate the old attr at the same time as when we do so on in labs/ssr.

Reply:
"launch" is a loose term here. Does this PR put the package in releasable state? Chrome 111 goes beta next well and stable in 4 weeks. Safari TP will be promoted relatively soon.

Either way is ultimately fine, though I do want to remove the old attribute as soon as possib.e.

Reply:
This PR does put the package in a releasable state. I think removing the old attribute in sync with core labs/ssr makes sense.